File: core/LongestIncreasingSubsequence/LongestIncreasingSubsequence.java
package com.dbc;

import java.util.Arrays;

public class LongestIncreasingSubsequence {
    public int lengthOfLIS(int[] nums) {
        int res = 1;
        int[] dp = new int[nums.length];
        Arrays.fill(dp, 1);
        for (int i = 0; i < nums.length; i++) {
            for (int j = 0; j < i; j++) {
                if (nums[j] < nums[i] && dp[j] >= dp[i]) {
                    dp[i] = dp[j] + 1;
                    res = Math.max(res, dp[i]);
                }
            }
        }
        return res;
    }
}

File: core/LongestIncreasingSubsequence/LongestIncreasingSubsequence.py
from typing import List

def lengthOfLIS(nums: List[int]) -> int:
    dp = [1] * len(nums)
    def search_index(start, end, num):
        remain = 0
        for i in range(start, end):
            if nums[i] < num and dp[i] > remain:
                remain = dp[i]
        return remain

    res = 0
    for index, item in enumerate(nums):
        dp[index] = search_index(0, index, item) + 1
        res = max(res, dp[index])

    return res
